langchain4j-vespa/src/test/java/dev/langchain4j/store/embedding/vespa/VespaEmbeddingStoreIT.java

```diff
@@ -0,0 +1,180 @@
+        Embedding embedding1 = embeddingModel.embed(segment1).content();
+
+        TextSegment segment2 = TextSegment.from("I've never been to New York.");
+        Embedding embedding2 = embeddingModel.embed(segment2).content();
+
+        TextSegment segment3 = TextSegment.from(
+            "But actually we tried our new swimming pool yesterday and it was awesome!"
+        );
+        Embedding embedding3 = embeddingModel.embed(segment3).content();
+
+        TextSegment segment4 = TextSegment.from("John Lennon was a very cool person.");
+        Embedding embedding4 = embeddingModel.embed(segment4).content();
+
+        embeddingStore.addAll(
+            asList(embedding1, embedding2, embedding3, embedding4),
+            asList(segment1, segment2, segment3, segment4)
+        );
+
+        // when
+        Embedding sportEmbedding = embeddingModel.embed("What is your favorite sport?").content();
+        List<EmbeddingMatch<TextSegment>> sportMatches = embeddingStore
+            .search(EmbeddingSearchRequest.builder().queryEmbedding(sportEmbedding).maxResults(2).build())
+            .matches();
+        Embedding musicEmbedding = embeddingModel.embed("And what about musicians?").content();
+        List<EmbeddingMatch<TextSegment>> musicMatches = embeddingStore
+            .search(EmbeddingSearchRequest.builder().queryEmbedding(musicEmbedding).maxResults(5).minScore(0.3).build())
+            .matches();
+
+        // then
+        assertThat(sportMatches).hasSize(2);
+        assertThat(sportMatches.get(0).score()).isCloseTo(0.639, percentage());
+        assertThat(sportMatches.get(0).embedded().text()).contains("football");
+        assertThat(sportMatches.get(1).score()).isCloseTo(0.232, percentage());
+        assertThat(sportMatches.get(1).embedded().text()).contains("swimming pool");
+
+        assertThat(musicMatches).hasSize(1);
+        assertThat(musicMatches.get(0).score()).isCloseTo(0.359, percentage());
+        assertThat(musicMatches.get(0).embedded().text()).contains("John Lennon");
+    }
+
+    @Override
+    protected EmbeddingStore<TextSegment> embeddingStore() {
+        return embeddingStore;
+    }
+
+    @Override
+    protected EmbeddingModel embeddingModel() {
+        return embeddingModel;
+    }
+
+    @Override
+    protected void clearStore() {
+        embeddingStore().removeAll();
+    }
+
+    @Override
+    protected void ensureStoreIsEmpty() {
+        await().atMost(Duration.ofSeconds(5)).untilAsserted(super::ensureStoreIsEmpty);
+    }
+
+    @Override
+    protected void ensureStoreIsReady() {
+        try {
+            getAllEmbeddings();
+        } catch (Exception e) {
+            deployVespaApp();
+        }
+    }
+
+    @Override
+    protected String documentURI(String id) {
+        return String.format("id:%s:%s::%s", DEFAULT_NAMESPACE, DEFAULT_DOCUMENT_TYPE, id);
+    }
+
+    @Override
+    protected Percentage percentage() {
+        return withPercentage(12);
```

Comment:
Yes, here I need a support if you think 12 is too much 😄 because less than 12 leads to assert failures. Since Vespa is highly configurable, could I may be adjust here anything to reduce percentage setting?
```
  rank-profile cosine_similarity {
    inputs {
      query(q) tensor<float>(x[384])
    }
    function cosine() {
      expression: 1.0 - distance(field, vector)
    }
    first-phase {
      expression: if (cosine() < query(threshold), -1, cosine())
      rank-score-drop-limit: 0.0
    }
  }
```

Reply:
I need to dig deeper into this, but from the surface 12% sounds very fishy, it might even hide a real problem when wrong embedding is matched...

Reply:
yes, if you think, it's fishy, any suggestions are welcome 🙏

Reply:
I'll try to look into it during this week!

Reply:
Look at schema/langchain4j.sd schema within vespa_app.zip - the Vespa config to play with is this one. I've just tried various variants but unfortunately without luck - the better difference I got - 7%. I suppose to be able to properly fix it I need to have good Vector DB knowledge and this is what I don't have currently :(

Reply:
Sorry, I did not have time to look into this yet :(

But the problem could be that Vespa returns cosine similarity (in the range of -1 to 1), while `EmbeddingMatch.score` should be a relevance score (in the range of 0 to 1). Cosine similarity can be converted into the relevance score using the `RelevanceScore.fromCosineSimilarity()` helper. There is a `CosineSimilarity.fromRelevanceScore()` as well BTW.

Reply:
good news to the end of the year 😄 got it working, this is the main change:
```
function relevance() {
  expression: ( cos(distance(field, vector)) + 1 ) / 2
}
```
in fact, previous config Vespa was returning the cosine similarity, now it returns the relevance score, exactly what the test expects.